nixos/modules/services/databases/neo4j.nix

```diff
@@ -36,114 +89,541 @@ let
+          data. Refer to the operations manual section on the
+          <link xlink:href="https://neo4j.com/docs/operations-manual/current/configuration/usage-data-collector/">Usage Data Collector</link>
+          for more information.
+        '';
+      };
     };
+
   };
 
   ###### implementation
 
-  config = mkIf cfg.enable {
-    systemd.services.neo4j = {
-      description = "Neo4j Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      environment = {
-        NEO4J_HOME = "${cfg.package}/share/neo4j";
-        NEO4J_CONF = "${cfg.dataDir}/conf";
-      };
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/neo4j console";
-        User = "neo4j";
-        PermissionsStartOnly = true;
-        LimitNOFILE = 40000;
-      };
-      preStart = ''
-        mkdir -m 0700 -p ${cfg.dataDir}/{data/graph.db,conf,logs}
-        ln -fs ${serverConfig} ${cfg.dataDir}/conf/neo4j.conf
-        if [ "$(id -u)" = 0 ]; then chown -R neo4j ${cfg.dataDir}; fi
-      '';
-    };
+  config =
+    let
+      policyNameList = (attrNames cfg.ssl.policies);
+      validPolicyNameList = ["legacy"] ++ policyNameList;
+      validPolicyNameString = concatStringsSep ", " validPolicyNameList;
+    in
+
+    mkIf cfg.enable {
+      assertions = [
+        { assertion = !(elem "legacy" policyNameList);
+          message = "The policy 'legacy' is special to Neo4j, and its name is reserved."; }
+        { assertion = elem cfg.bolt.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.bolt.sslPolicy = ${cfg.bolt.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+        { assertion = elem cfg.https.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.https.sslPolicy = ${cfg.https.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+      ];
+
+      systemd.services.neo4j = {
+        description = "Neo4j Daemon";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        environment = {
+          NEO4J_HOME = "${cfg.package}/share/neo4j";
+          NEO4J_CONF = "${toString cfg.directories.home}/conf";
+        };
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/neo4j console";
+          User = "neo4j";
+          PermissionsStartOnly = true;
+          LimitNOFILE = 40000;
+        };
+
+        preStart = ''
+          # Directories Setup
+          mkdir -m 0700 -p ${toString cfg.directories.home}/{certificates,conf,data,import,logs,plugins}
 
-    environment.systemPackages = [ cfg.package ];
+          # SSL Policies Directory setup
+          #   Only generate required directories when default certificates and baseDirectory is used.
+          ${concatStringsSep "\n" (mapAttrsToList (name: value:
+            if cfg.directories.home + "/certificates/" + name == value.baseDirectory then
```

Comment:
Shouldn't this be the policies directory?

Also, you can use the same `highestPrio` thing here to not have to even do this string comparison.

Reply:
> you can use the same `highestPrio` thing here

How does one access an attribute within the policy without knowing its name?
`options.services.neo4j.ssl.policies.?.baseDirectory.highestPrio`

Reply:
filter `builtins.attrValues options.services.neo4j.ssl.policies` based on `(opt: opt.baseDirectory.highestPrio < 1500)`, then an `if` on whether there are any results at all, then a `concatMapStringsSep " "` over `(opt: opt.baseDirectory.value)`.

Reply:
> filter `builtins.attrValues options.services.neo4j.ssl.policies` based on `(opt: opt.baseDirectory.highestPrio < 1500)`

This doesn't work because `options.services.neo4j.ssl.policies` is an option itself, and the `attrValues` of `options.services.neo4j.ssl.policies` are not options as you expected. A trace of the `attrNames` is:
```
[ "_type" "declarations" "default" "definitions" "description" "files" "highestPrio" "isDefined" "loc" "options" "type" "value" ]
```
The `options` attribute of this set sounded promising, but alas is empty.

I attempted to access options of a specific defined policy by first collecting the policy names, which also fails, i.e. `options.services.neo4j.ssl.policies."name"` results is `attribute 'name' missing`, even when the 'name' policy is defined.

The current push incorrectly uses the `definitions` attribute, with the assumption that any option not defined therein must still be at its default. _Incorrectly_ in that it doesn't merge all attrsets in `definitions` before checking for this assumption (_naively chooses the first definition attrset_). 

Of course it's Nixos' job to merge these definitions, but I still cannot figure out how/where to access a policy's option attrset.